.github/workflows/test.yml

```diff
@@ -0,0 +1,66 @@
+name: tests
+
+on: [push, pull_request]
+
+jobs:
+  code-analysis:
+    runs-on: ubuntu-latest
+    steps:
```

Comment:
The first two steps are run for each of the jobs. One way to optimize would be running all of these in a matrix and breaking out the steps that are shared. It would make this file look a lot cleaner.

Reply:
Not sure I agree that the matrix approach is cleaner. Would be less code, but also less intuitive. Perhaps just my inexperience talking... Also, 1 of the 3 jobs have a unique step which doesn't make sense for the other two jobs. I was looking into using composite actions for this, but seemed to just add more complexity in another way.

Unless you have strong opinions on this, I'll keep this as is.